parquet/src/main/java/org/apache/iceberg/parquet/ParquetFormatModel.java

```diff
@@ -0,0 +1,394 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.parquet;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.iceberg.FileContent;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.MetricsConfig;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.data.parquet.GenericParquetWriter;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.formats.FormatModel;
+import org.apache.iceberg.formats.ReadBuilder;
+import org.apache.iceberg.formats.WriteBuilder;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.DeleteSchemaUtil;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.mapping.NameMapping;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.parquet.column.ParquetProperties;
+import org.apache.parquet.schema.MessageType;
+
+public class ParquetFormatModel<D, S, F> implements FormatModel<D, S> {
+  public static final String WRITER_VERSION_KEY = "parquet.writer.version";
+
+  private final Class<? extends D> registerType;
+  private final Class<S> schemaType;
+  private final ReaderFunction<D> readerFunction;
+  private final BatchReaderFunction<D, F> batchReaderFunction;
+  private final WriterFunction<S> writerFunction;
+
+  private ParquetFormatModel(
+      Class<? extends D> registerType,
+      Class<S> schemaType,
+      ReaderFunction<D> readerFunction,
+      BatchReaderFunction<D, F> batchReaderFunction,
+      WriterFunction<S> writerFunction) {
+    this.registerType = registerType;
+    this.schemaType = schemaType;
+    this.readerFunction = readerFunction;
+    this.batchReaderFunction = batchReaderFunction;
+    this.writerFunction = writerFunction;
+  }
+
+  public ParquetFormatModel(Class<D> type) {
+    this(type, null, null, null, null);
+  }
+
+  public ParquetFormatModel(
+      Class<D> type,
+      Class<S> schemaType,
+      ReaderFunction<D> readerFunction,
+      WriterFunction<S> writerFunction) {
+    this(type, schemaType, readerFunction, null, writerFunction);
+  }
+
+  public ParquetFormatModel(
+      Class<? extends D> returnType,
+      Class<S> schemaType,
+      BatchReaderFunction<D, F> batchReaderFunction) {
+    this(returnType, schemaType, null, (BatchReaderFunction<D, F>) batchReaderFunction, null);
+  }
+
+  @Override
+  public FileFormat format() {
+    return FileFormat.PARQUET;
+  }
+
+  @Override
+  public Class<D> type() {
+    return (Class<D>) registerType;
```

Comment:
Unchecked cast.

Reply:
The cast was ok as the `registerType` was defined like this:
```
private final Class<? extends D> registerType;
```
Depends on https://github.com/apache/iceberg/pull/12298#discussion_r2548086267